Source/Protected/ProtectedAsMock.cs

```diff
@@ -0,0 +1,231 @@
+		{
+			return Mock.Setup(this.mock, (Expression<Func<T, TResult>>)ReplaceDuck(expression), null);
+		}
+
+		private static LambdaExpression ReplaceDuck(LambdaExpression expression)
+		{
+			var targetParameter = Expression.Parameter(typeof(T), expression.Parameters[0].Name);
+			return Expression.Lambda(DuckReplacerInstance.Visit(expression.Body), targetParameter);
+		}
+
+		/// <summary>
+		/// <see cref="ExpressionVisitor"/> used to replace occurrences of `TDuck.Member` sub-expressions with `T.Member`.
+		/// </summary>
+		private sealed class DuckReplacer : ExpressionVisitor
+		{
+			private Type duckType;
+			private Type targetType;
+
+			public DuckReplacer(Type duckType, Type targetType)
+			{
+				this.duckType = duckType;
+				this.targetType = targetType;
+			}
+
+			protected override Expression VisitMethodCall(MethodCallExpression node)
+			{
+				if (node.Object is ParameterExpression left && left.Type == this.duckType)
+				{
+					var targetParameter = Expression.Parameter(typeof(T), left.Name);
+					return Expression.Call(targetParameter, FindCorrespondingMethod(node.Method), node.Arguments);
+				}
+				else
+				{
+					return node;
+				}
+			}
+
+			protected override Expression VisitMember(MemberExpression node)
+			{
+				if (node.Expression is ParameterExpression left && left.Type == this.duckType)
+				{
+					var targetParameter = Expression.Parameter(typeof(T), left.Name);
+					return Expression.MakeMemberAccess(targetParameter, FindCorrespondingMember(node.Member));
+				}
+				else
+				{
+					return node;
+				}
+			}
+
+			private MemberInfo FindCorrespondingMember(MemberInfo duckMember)
+			{
+				if (duckMember is MethodInfo duckMethod)
+				{
+					return FindCorrespondingMethod(duckMethod);
+				}
+				else if (duckMember is PropertyInfo duckProperty)
+				{
+					return FindCorrespondingProperty(duckProperty);
+				}
+				else
+				{
+					throw new NotSupportedException();
+				}
+			}
+
+			private MethodInfo FindCorrespondingMethod(MethodInfo duckMethod)
+			{
+				var candidateTargetMethods =
+					this.targetType
+				    .GetMethods(BindingFlags.NonPublic | BindingFlags.Instance)
+				    .Where(ctm => IsCorrespondingMethod(duckMethod, ctm));
+
+				var targetMethod = candidateTargetMethods.Single();
+
+				if (targetMethod.IsGenericMethodDefinition)
+				{
+					var duckGenericArgs = duckMethod.GetGenericArguments();
+					targetMethod = targetMethod.MakeGenericMethod(duckGenericArgs);
+				}
+
+				return targetMethod;
+			}
+
+			private PropertyInfo FindCorrespondingProperty(PropertyInfo duckProperty)
+			{
+				var candidateTargetProperties =
+					this.targetType
+					.GetProperties(BindingFlags.NonPublic | BindingFlags.Instance)
+					.Where(ctp => IsCorrespondingProperty(duckProperty, ctp));
+
+				return candidateTargetProperties.Single();
+			}
+
+			private static bool IsCorrespondingMethod(MethodInfo duckMethod, MethodInfo candidateTargetMethod)
```

Comment:
Check if we are comparing method signatures for a precise match anywhere else in Moq's code base, and either reuse existing code, or make this implementation here reusable. Also check if the framework already offers something to compare method signatures&mdash;no need to reinvent the wheel.